### PR TITLE
CSV previews are sometimes broken - refactoring of CSV previews controller

### PR DESCRIPTION
**Pull request by unoduetre**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

This is a refactoring related to the previous work on CSV previews: https://github.com/alphagov/frontend/pull/3945

## Why

[Trello card](https://trello.com/c/iaw1zR77/2328-csv-previews-are-sometimes-broken-l), [Jira issue NAV-12163](https://gov-uk.atlassian.net/browse/NAV-12163)

### CODE REVIEW

File: test/unit/services/csv_preview_service_test.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 require "test_helper"
 
-class CsvPreviewControllerTest < ActionController::TestCase
+class CsvPreviewServiceTest < ActionController::TestCase
```
Comment (hannako) [inline test/unit/services/csv_preview_service_test.rb:3]: This is a bit of a confused spec...see docs here for what the [ActionController::TestCase]( https://api.rubyonrails.org/classes/ActionController/TestCase.html) is meant to test. It also seems to be testing all the private methods, which again is not conventional (or particularly informative)

I think in this PR we should at a minimum implement a unit test for the `CSVPreviewService` that only tests the public `csv_rows` method. We can involve the same csv data types to make sure we aren't losing coverage.

Stretch goal:

- a `CsvPreviewServiceControllerTest < ActionController::TestCase ` test, that is actually a controller test. Currently we don't have a controller test at all, just a massive [integration test](https://github.com/alphagov/frontend/blob/main/test/integration/csv_preview_test.rb). Probably the easiest thing to do is to [cherry pick from that integration test](https://github.com/alphagov/frontend/blob/main/test/integration/csv_preview_test.rb) . But this is a nice to have, as we do have test coverage in the integration test. So we could do this later/in a separate piece of work if you don't want to do it now.

**Review by hannako — COMMENTED** [review]
Thanks for doing this Mat, I think this is a great improvement. I've made a couple of notes, mainly around the spec.

File: test/unit/services/csv_preview_service_test.rb (hunk 1 above)
Comment (unoduetre) [reply test/unit/services/csv_preview_service_test.rb:3]: Argh, I missed that.

Comment (unoduetre) [reply test/unit/services/csv_preview_service_test.rb:3]: OK, I'll add a more complete unit test for the CSVPreviewService.

File: app/services/csv_preview_service.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
+class CsvPreviewService
+  MAXIMUM_COLUMNS = 50
+  MAXIMUM_ROWS = 1000
+
+  def initialize(id:, filename:)
+    @id = id
+    @filename = filename
```
[the remark this answers is not in the document]
Comment (unoduetre) [reply app/services/csv_preview_service.rb:7]: Yeah, it's a good idea. I will change that.

[the remark this answers is not in the document]
Comment (unoduetre) [reply app/services/csv_preview_service.rb]: Changed.

File: test/unit/services/csv_preview_service_test.rb (hunk 1 above)
Comment (unoduetre) [reply test/unit/services/csv_preview_service_test.rb]: Added.

File: app/controllers/csv_preview_controller.rb [hunk 3]
```diff
@@ -25,35 +23,20 @@ def show
...
+private
+
   def access_limited
-    render :access_limited, status: :forbidden and return
+    render :access_limited, status: :forbidden
   end
 
-private
+  def malformed_csv
+    render :malformed_csv
```
Comment (hannako) [inline app/controllers/csv_preview_controller.rb:38]: How come we no longer need the `and return` ?

Comment (hannako) [reply app/controllers/csv_preview_controller.rb:38]: As there's no controller test, I can't tell if this is an issue or not.

Comment (unoduetre) [reply app/controllers/csv_preview_controller.rb:38]: return returns from the function, so it's redundant. What do you think return does in this code?

The controller has originally only had the integration test. I can add tests for that to it, or I can create a functional test and add it there. As this task was a refactoring task and not "write missing tests" task, I've only added tests which relate to the moved code, but I can add the missing controller tests, if that's your preference.

Comment (hannako) [reply app/controllers/csv_preview_controller.rb:38]: I was thinking without the return we might get https://api.rubyonrails.org/classes/AbstractController/DoubleRenderError.html 

But if that's not then happy to lose it. Thanks for adding the test.

Comment (unoduetre) [reply app/controllers/csv_preview_controller.rb:38]: I've added the test for malformed csv.

Comment (unoduetre) [reply app/controllers/csv_preview_controller.rb:38]: https://github.com/alphagov/frontend/blob/8c08ee54269049bbf530846e1e4f8cf6e632073e/test/integration/csv_preview_test.rb#L112

**Review by hannako — APPROVED** [review]
Looks great, thanks Mat

**Pull request merged**